js/src/snapshot.ts

```diff
@@ -0,0 +1,453 @@
+        ['fee_validation_st_sol_total', 'u64'],
+        ['fee_developer_st_sol_total', 'u64'],
+        ['deposit_amount', LamportsHistogram],
+        ['withdraw_amount', WithdrawMetric],
+      ],
+    },
+  ],
+  [
+    SeedRange,
+    {
+      kind: 'struct',
+      fields: [
+        ['begin', 'u64'],
+        ['end', 'u64'],
+      ],
+    },
+  ],
+  [
+    Validator,
+    {
+      kind: 'struct',
+      fields: [
+        ['fee_credit', 'u64'],
+        ['fee_address', 'u256'],
+        ['stake_seeds', SeedRange],
+        ['unstake_seeds', SeedRange],
+        ['stake_accounts_balance', 'u64'],
+        ['unstake_accounts_balance', 'u64'],
+        ['active', 'u8'],
+      ],
+    },
+  ],
+  [
+    ValidatorPubKeyAndEntry,
+    {
+      kind: 'struct',
+      fields: [
+        ['pubkey', 'u256'],
+        ['entry', Validator],
+      ],
+    },
+  ],
+  [
+    MaintainerPubKeyAndEntry,
+    {
+      kind: 'struct',
+      fields: [
+        ['pubkey', 'u256'],
+        ['entry', [0]],
+      ],
+    },
+  ],
+  [
+    RewardDistribution,
+    {
+      kind: 'struct',
+      fields: [
+        ['treasury_fee', 'u32'],
+        ['validation_fee', 'u32'],
+        ['developer_fee', 'u32'],
+        ['st_sol_appreciation', 'u32'],
+      ],
+    },
+  ],
+  [
+    FeeRecipients,
+    {
+      kind: 'struct',
+      fields: [
+        ['treasury_account', 'u256'],
+        ['developer_account', 'u256'],
+      ],
+    },
+  ],
+  [
+    Validators,
+    {
+      kind: 'struct',
+      fields: [
+        ['entries', [ValidatorPubKeyAndEntry]],
+        ['maximum_entries', 'u32'],
+      ],
+    },
+  ],
+  [
+    Maintainers,
+    {
+      kind: 'struct',
+      fields: [
+        ['entries', [MaintainerPubKeyAndEntry]],
+        ['maximum_entries', 'u32'],
+      ],
+    },
+  ],
+  [
+    Solido,
+    {
+      kind: 'struct',
+      fields: [
+        ['lido_version', 'u8'],
+
+        ['manager', 'u256'],
+
+        ['st_sol_mint', 'u256'],
+
+        ['exchange_rate', ExchangeRate],
+
+        ['sol_reserve_authority_bump_seed', 'u8'],
+        ['stake_authority_bump_seed', 'u8'],
+        ['mint_authority_bump_seed', 'u8'],
+        ['rewards_withdraw_authority_bump_seed', 'u8'],
+
+        ['reward_distribution', RewardDistribution],
+
+        ['fee_recipients', FeeRecipients],
+
+        ['metrics', Metrics],
+
+        ['validators', Validators],
+
+        ['maintainers', Maintainers],
+      ],
+    },
+  ],
+]);
+
+export const getSolido = (solidoInstanceDataBuffer: Buffer) => {
+  const deserialized = deserializeUnchecked(
+    schema,
+    Solido,
+    solidoInstanceDataBuffer
+  );
+
+  return deserialized;
+};
+
+export const getSnapshot = async (
+  connection: Connection,
+  solidoInstanceAccountInfo: AccountInfo<Buffer>,
+  reserveAccountInfo: AccountInfo<Buffer>,
+  programAddresses: ProgramAddresses
+): Promise<Snapshot> => {
+  const solido = getSolido(solidoInstanceAccountInfo.data);
```

Comment:
Now we still do a torn read, the `getMultipleAccounts` later on can read from a later slot than where we got the `solido` account from, and they may have changed in the meantime. To get a consistent view, we have to re-read the `solido` account as part of the `getMultipleAccounts`. One way to do this is in a loop:

* Start with just the Solido account, do a `getMultipleAccounts`, put the results in a dictionary keyed on accounts.
* Start parsing the Solido state and the other accounts, but instead of calling the RPC, read from the dictionary.
* If an account is not present in the dictionary, add that address to the list of accounts to retrieve, and start again. To be more efficient, we can add all validator stake accounts to the list immediately. Then we should only have to do two reads (the first one to get the Solido instance, and the second one that gets everything we need). In the extremely unlucky case where you run this just as we add a validator, it will have to do three iterations.